### DISCUSSION TITLE
Method pattern asynchronous

### DISCUSSION THREAD

**Discussion by StefanMaron** [microsoft/alguidelines#184] [BC Best Practices]
I have just implemented a additional function on the table in order to have the method not run in the client session but instead, have it run in by the TaskScheduler. 

``` AL
procedure UpdateQuantity()
var
    UpdateQuantityMethPTE: Codeunit UpdateQuantityMethPTE;
begin
    UpdateQuantityMethPTE.UpdateItemQuantity(Rec);
end;

procedure UpdateQuantityAsync()
begin
    TaskScheduler.CreateTask(Codeunit::UpdateQuantityMethPTE, 0, true, CompanyName, CurrentDateTime, Rec.RecordId);
end;
```

To make this work, the Method codeunit needs to be based on the table of course

``` AL
codeunit 71100 "UpdateQuantityMethPTE"
{
    TableNo = ItemTotalQuantityPTE;

    trigger OnRun()
    begin
        UpdateItemQuantity(Rec);
    end;

    internal procedure UpdateItemQuantity(var ItemTotalQuantity: Record ItemTotalQuantityPTE)
    var
        IsHandled: Boolean;
    begin
    ....
```

What do you think? Could be a nice addition to the Method Pattern by @waldo1001

**Comment by waldo1001** [comment]
I like it!  I like it a lot, actually 🤔

**Reply by StefanMaron** [reply]
the only problem I currently see is, that the codeunit can be run directly instead of using the procedure at the table. The intent was to make the codeunit not callable directly by setting the main function to internal. 
For the onrun, the only way to achieve the same is, to make the whole codeunit internal. But that would also hide the event publishers

**Reply by waldo1001** [reply]
Can't we move the Async-function in the codeunit - and let it run itself? 🤔

**Reply by StefanMaron** [reply]
Not sure if I understand. CreateTask only takes a codeunit as paramter. Does not work without the OnRun()

**Reply by waldo1001** [reply]
Something like:

```al
codeunit 60200 "SomeName Meth"
{
    TableNo = customer;

    trigger OnRun()
    begin
        SomeName(Rec);
    end;

    internal procedure SomeName(var Rec: Record Customer)
    var
        IsHandled: Boolean;
    begin
        OnBeforeSomeName(Rec, IsHandled);

        DoSomeName(Rec, IsHandled);

        OnAfterSomeName(Rec);
    end;

    internal procedure SomeNameAsync(var Rec: Record Customer)
    begin
        TaskScheduler.CreateTask(Codeunit::"SomeName Meth", 0, true, CompanyName, CurrentDateTime, Rec.RecordId);
    end;

    local procedure DoSomeName(var Rec: Record Customer; IsHandled: Boolean);
    begin
        if IsHandled then
            exit;

    end;
...
```

**Reply by StefanMaron** [reply]
Yes, of course. That will work as well. But the reason for the internal procedure was, that the entry point to the method should be table. The codeunit should not be used directly. I mean by depending apps. By defining a OnRun() trigger, the codeunit can be accessed directly.

Not sure though if this really is a problem to worry about 🤔

**Reply by StefanMaron** [reply]
Important is, that the internal procedure `SomeName` can not have the same name as the procedure on the table. This will lead to implicit with in the onRun trigger

**Reply by waldo1001** [reply]
Hm - good point.  It should have a different parameterset, no?

**Reply by StefanMaron** [reply]
On the table there is no parameter, in there codeunit the record is the paramter. I got an error saying that the parameter is not valid. It seems to only pick up the procedure from the table and ignores the procedure from the codeunit/local scope

A rename of the codeunit procedure did the job

**Reply by waldo1001** [reply]
Sounds like a bug :-/.

**Reply by StefanMaron** [reply]
![image](https://user-images.githubusercontent.com/24838311/169991875-e15c1bff-d45f-49cb-a26d-5f4bd57c6f43.png)

But this can be fixed easily by adding `NoImplicitWith`
![image](https://user-images.githubusercontent.com/24838311/169992043-9cebd3b8-bf53-4058-b61c-ca3f1fc8fe2b.png)

![image](https://user-images.githubusercontent.com/24838311/169992003-49bb4fb2-c9de-4493-9dad-2652727a7ab5.png)

**Comment by StefanMaron** [comment]
For later reference:

This is also a good idea I think:
![image](https://user-images.githubusercontent.com/24838311/170203208-55d734d1-f969-4aaa-97b7-594dce35bab6.png)

https://twitter.com/ajkauffmann/status/1529084189198491648

**Comment by StefanMaron** [comment]
And one more addition. This is the motivation behind and the why for doing asynch more often:
https://thedoubleh.dev/posts/performancevsexperience/

post by @henrikwh
